src/Protocol/ActivityPub.php

```diff
@@ -0,0 +1,1820 @@
+		if (($item['gravity'] != GRAVITY_PARENT) && !Item::exists(['uri' => $item['parent-uri']])) {
+			logger('Parent ' . $item['parent-uri'] . ' not found, message will be discarded.', LOGGER_DEBUG);
+			return;
+		}
+
+		$item['network'] = Protocol::ACTIVITYPUB;
+		$item['private'] = !in_array(0, $activity['receiver']);
+		$item['author-id'] = Contact::getIdForURL($activity['author'], 0, true);
+		$item['owner-id'] = Contact::getIdForURL($activity['owner'], 0, true);
+		$item['uri'] = $activity['id'];
+		$item['created'] = $activity['published'];
+		$item['edited'] = $activity['updated'];
+		$item['guid'] = $activity['diaspora:guid'];
+		$item['title'] = HTML::toBBCode($activity['name']);
+		$item['content-warning'] = HTML::toBBCode($activity['summary']);
+		$item['body'] = self::convertMentions(HTML::toBBCode($activity['content']));
+		$item['location'] = $activity['location'];
+		$item['tag'] = self::constructTagList($activity['tags'], $activity['sensitive']);
+		$item['app'] = $activity['service'];
+		$item['plink'] = defaults($activity, 'alternate-url', $item['uri']);
+
+		$item = self::constructAttachList($activity['attachments'], $item);
+
+		$source = JsonLD::fetchElement($activity, 'source', 'content', 'mediaType', 'text/bbcode');
+		if (!empty($source)) {
+			$item['body'] = $source;
+		}
+
+		$item['protocol'] = Conversation::PARCEL_ACTIVITYPUB;
+		$item['source'] = $body;
+		$item['conversation-href'] = $activity['context'];
+		$item['conversation-uri'] = $activity['conversation'];
+
+		foreach ($activity['receiver'] as $receiver) {
+			$item['uid'] = $receiver;
+			$item['contact-id'] = Contact::getIdForURL($activity['author'], $receiver, true);
+
+			if (($receiver != 0) && empty($item['contact-id'])) {
+				$item['contact-id'] = Contact::getIdForURL($activity['author'], 0, true);
+			}
+
+			$item_id = Item::insert($item);
+			logger('Storing for user ' . $item['uid'] . ': ' . $item_id);
+		}
+	}
+
+	/**
+	 * @brief 
+	 *
+	 * @param $url
+	 * @param $child
+	 */
+	private static function fetchMissingActivity($url, $child)
+	{
+		if (Config::get('system', 'ostatus_full_threads')) {
+			return;
+		}
+
+		$object = ActivityPub::fetchContent($url);
+		if (empty($object)) {
+			logger('Activity ' . $url . ' was not fetchable, aborting.');
+			return;
+		}
+
+		$activity = [];
+		$activity['@context'] = $object['@context'];
+		unset($object['@context']);
+		$activity['id'] = $object['id'];
+		$activity['to'] = defaults($object, 'to', []);
+		$activity['cc'] = defaults($object, 'cc', []);
+		$activity['actor'] = $child['author'];
+		$activity['object'] = $object;
+		$activity['published'] = $object['published'];
+		$activity['type'] = 'Create';
+
+		self::processActivity($activity);
+		logger('Activity ' . $url . ' had been fetched and processed.');
+	}
+
+	/**
+	 * @brief Returns the user id of a given profile url
+	 *
+	 * @param string $profile
+	 *
+	 * @return integer user id
+	 */
+	private static function getUserOfProfile($profile)
```

Comment:
wouldn't it be better to have `$url` instead of `$profile` as parameter?

Reply:
It is the profile url. So possibly $profile is better than a generic $url? I had the same idea.

Reply:
This method queries the `contact` table, so it probably should be something like `Model\Contact::getSelfUserIdByUrl($url)`.

Reply:
Possibly this should move into the ```User``` class since this handles the user-id - although it does this with the contact table.

Reply:
Possibly ```User::getUIDByProfileLink(...)```?